spec/models/zone_spec.rb

```diff
@@ -94,6 +94,23 @@
     end
   end
 
+  describe "#ems_metrics_collectable" do
+    let(:zone)                       { FactoryBot.create(:zone) }
+    let(:ems_supporting_metrics)     { FactoryBot.create(:ems_infra, :zone => zone) }
+    let(:ems_not_supporting_metrics) { FactoryBot.create(:ems_cloud, :zone => zone) }
+
+    before do
+      allow_any_instance_of(ExtManagementSystem).to receive(:supports?).with(:metrics).and_return(true)
```

Comment:
This is not ideal, but now that we're doing instance-level checks for `supports?(:metrics)` we either have to do this, wait for provider PRs to be merged that add `supports :metrics` to the respective classes (VMware and Amazon in this case back the "generic" :ems_infra and :ems_cloud factories), or try to stub the AR query and return these objects that we can then stub the `:supports?` method for.

Reply:
Yeah I think this could be a great candidate for pluggable schedules that we talked about somewhere a while ago.

Each provider which supports metrics capture can bring their own version of `metric_capture_perf_capture_timer` so core doesn't need to do `MiqServer.my_server.zone.ems_metrics_collectable.each` at all.

Reply:
Introducing a metrics helper class for the ems that support metrics. It would be responsible for submitting a message (well, a plethora of messages) on the queue.

Would probably be best if the vendor specific ems class registered the timer, but not sure the best short term solution

Reply:
Provider PRs are merged now...not sure if you want to undo or leave as is.